indra/newview/llvovolume.cpp

```diff
@@ -832,12 +841,25 @@ void LLVOVolume::updateTextureVirtualSize(bool forced)
         // if the face has gotten small enough to turn off texture animation and texture
         // animation is running, rebuild the render batch for this face to turn off
         // texture animation
+        // Do the opposite when the face gets big enough.
+        // If a face is animatable, it will always have non-null mTextureMatrix
+        // pointer defined after the first call to LLVOVolume::animateTextures,
+        // although the animation is not always turned on.
         if (face->mTextureMatrix != NULL)
         {
-            if ((vsize < MIN_TEX_ANIM_SIZE && old_size > MIN_TEX_ANIM_SIZE) ||
-                (vsize > MIN_TEX_ANIM_SIZE && old_size < MIN_TEX_ANIM_SIZE))
+            if ((vsize > MIN_TEX_ANIM_SIZE) != (old_size > MIN_TEX_ANIM_SIZE))
             {
                 gPipeline.markRebuild(mDrawable, LLDrawable::REBUILD_TCOORD);
+                // dirtyGeom+markRebuild tells the engine to call
+                // LLVolumeGeometryManager::rebuildGeom, which rebuilds the
+                // LLDrawInfo for the spatial group containing this LLFace,
+                // safely copying the mTextureMatrix from the LLFace the the
+                // LLDrawInfo. While it's not ideal to call it here, prims with
+                // animated faces get moved to a smaller partition to reduce
+                // side-effects of their updates (see shrinkWrap in
+                // LLVOVolume::animateTextures).
+                mDrawable->getSpatialGroup()->dirtyGeom();
+                gPipeline.markRebuild(mDrawable->getSpatialGroup());
```

Comment:
Main fix is here

Reply:
Is this logic going to work? If virtual size is below that threshold, then the matrix will be filled and it might be possible the rebuild won't happen later since the matrix would not be null.

What's the harm in removing the virtual size qualification?

Reply:
> What's the harm in removing the virtual size qualification?

The virtual size comparison logic has not changed, I've just cleaned it up.

For the rest of your questions, see also `LLVolumeGeometryManager::registerFace` for the logic that ultimately determines whether the texture matrix gets copied. The `dirtyGeom` and `markRebuild` lines I've added force that function to get recalled for faces impacted by this bug.

Reply:
My repository is behind (downstream Firestorm), so this is the first time I have seen the added virtual size qualification and I apologize if it was added by someone else.

But I think the combination of the two qualifications (null check on the matrix, virtual size threshold) narrows the chances of a rebuild down too far. Which is why I was asking if the second was even necessary.

If the matrix is null and virtual size is outside the threshold, then the rebuild will not happen until some other outside logic forces a rebuild because the first qualification will never pass again.

LLVolumeGeometryManager::registerFace seems to only be called when the face is added to the spatial group, so when will the face be rebuilt once the virtual size threshold is met? When the LOD on the drawable changes? What if it has only one LOD level?

I think the bug will persist if the face is distant or small and enters into frustum from a previous cull. In this situation and if the second qualification is used, I think it will not start animating as the camera get closer and the virtual size grows larger than the threshold until a LOD level change initiates a rebuild.

I know this is possibly a narrow set of circumstances but I think it will be noticeable when it does happen and I do not see any harm in removing the virtual size qualification.

Reply:
> If the matrix is null

This is explained in the comments, as well as the edge case fix that I also included in `LLVOVolume::animateTextures`

> LOD

I didn't change the LOD code

> I do not see any harm in removing the virtual size qualification

I already told you that code has not changed. Can you please confirm exactly which subset of the code diff you are referring to by copying and pasting it in your reply, and explain exactly how the code change maps to the issue you are describing?